datalogger/logger.go

```diff
@@ -95,3 +107,122 @@ func (logger *CSVDataLogger[T]) Close() bool {
 	logger.isOpen = false
 	return true
 }
+
+func CreateJSONDataLogger[T any](filename string) (DataLogger[T], error) {
+	fmt.Printf("Creating a JSON data logger: %v!\n", filename)
+	data := make([]T, 0)
+	destination, err := os.Create(filename)
+	if err != nil {
+		return &JSONDataLogger[T]{&sync.Mutex{}, 0, data, true, destination}, err
+	}
+
+	result := JSONDataLogger[T]{&sync.Mutex{}, 0, data, true, destination}
+	return &result, nil
+}
+
+func (logger *JSONDataLogger[T]) LogRecord(record T) {
+	logger.mut.Lock()
+	defer logger.mut.Unlock()
+	logger.recordCount += 1
+	logger.data = append(logger.data, record)
+}
+
+func PrettyPrint(out string, spaces int) string {
+	return fmt.Sprintf("%*s%s\n", spaces, "", out)
+}
+
+// Key Value Pair Print
+func KVPPrint(key string, value any) string {
+	return fmt.Sprintf("\"%s\": \"%v\"", key, value)
+}
+
+func (logger *JSONDataLogger[T]) Export() bool {
+	logger.mut.Lock()
+	defer logger.mut.Unlock()
+	if !logger.isOpen {
+		return false
+	}
+
+	// The rest of the code is written with the assumption that the indexing stays the same
+	visibleFields := reflect.VisibleFields(reflect.TypeOf((*T)(nil)).Elem())
+
+	// Meta Data
+	keys := make([]string, 0)
+	descriptions := make([]string, 0)
+	for _, v := range visibleFields {
+		// Keys
+		key, keysuccess := v.Tag.Lookup("json")
+		if !keysuccess {
+			key = fmt.Sprintf("None for %s", v.Name)
+		}
+		keys = append(keys, key)
+
+		// Descriptions
+		description, success := v.Tag.Lookup("Description")
+		if !success {
+			description = fmt.Sprintf("None for %s", v.Name)
+		}
+		descriptions = append(descriptions, description)
+	}
+
+	spaces := int(0)
+
+	// Open
+	logger.destination.Write([]byte(PrettyPrint("{", spaces)))
+	spaces += JSONSpaces
+
+	// Write Meta Data
+	logger.destination.Write([]byte(PrettyPrint("\"[METADATA]\": {", spaces)))
+	spaces += JSONSpaces
+	for i, key := range keys {
+		if i == len(keys)-1 {
+			logger.destination.Write([]byte(PrettyPrint(KVPPrint(key, descriptions[i]), spaces)))
+		} else {
+			logger.destination.Write([]byte(PrettyPrint(KVPPrint(key, descriptions[i])+",", spaces)))
+		}
+	}
+	spaces -= JSONSpaces
+	logger.destination.Write([]byte(PrettyPrint("},", spaces)))
+
+	// Write Data
+	logger.destination.Write([]byte(PrettyPrint("\"Data\": [", spaces)))
+	spaces += JSONSpaces
+	for ix, d := range logger.data {
+		logger.destination.Write([]byte(PrettyPrint("{", spaces)))
+		spaces += JSONSpaces
+		data := reflect.ValueOf(d)
+		for i, v := range visibleFields {
+			toWrite := data.FieldByIndex(v.Index)
```

Comment:
I think that you are safe here! (Going back to the question above). Using `v.Index` is different than using `i` and assuming that the fields are in the same order. I would *imagine* that the `Index` field is a special "thing" so that they are guaranteed to map. What do you think?

Reply:
I agree! Do you think that the comment is redundant then - specifically because it is a safe assumption?